File: lua/ingame/game_over.lua
-- lua/ingame/game_over.lua
-- Context: InGame (read-only). Backs NO catalog declaration by design -- see the parity note
-- below. Dispatched by src/civsim_harness/run/game_over.py's LuaGameOverReader, from
-- run/turn_cycle.py's pre-save position, once per turn.
--
-- WHY THIS FILE EXISTS (MEASURED 2026-09-21, 18:50 EDT, live Linux client 1.0.12.9):
-- Persia was defeated at game turn 59 (Georgia took the only city). The harness had no game-over
-- detection, so the next turn began as any other: the FR-007 start-of-turn quicksave was
-- requested, the already-finished game never wrote one, and the run paused with
-- `SaveVerificationError("expected .Civ6Save not found ... waited 10 s")`. A fresh InGame read
-- taken afterwards showed `Players[0]:IsAlive() == false` and `EndGameMenu` as the only shown
-- screen. The run's honest terminal state is a defeat; this file is what lets the harness say so
-- before it asks a finished game for a save it will never write.
--
-- PARITY BOUNDARY (Principle I, catalogs/README.md §3.5) -- READ BEFORE EXTENDING.
-- `lua/gamecore/game_outcome.lua` (T216, declaration `game.outcome_state`) reports whether the
-- LOCAL player has won or lost and deliberately reports NOTHING about who else won or how,
-- because while the game is still being played that is opponent state a human cannot see. That
-- boundary is unchanged and still binding for that file.
--
-- This file is the other side of the same line: the moment the game is over, Firaxis's own
-- end-game screen states all of it to the human in plain text -- the victory's own name and the
-- winning civilization -- so reading those two, and ONLY once the game is already over, is
-- human-parity-correct rather than a leak. The gate is structural below: `victory_type` and
-- `winner` are computed only after `game_over` has already been established, and are absent from
-- every other result. Do NOT hoist them above that gate, and do NOT extend this file with
-- `Game.GetVictoryProgressForTeam(...)`, another player's score, or any other civ's alive/dead
-- state -- none of those is on the end-game screen.
--
-- This read is harness bookkeeping at a turn boundary, not part of the agent's observation
-- surface: it carries no `ParityDeclaration` and never reaches a decision request. What the agent
-- sees about outcomes remains exactly `game.outcome_state`.
--
-- WHAT IT MIRRORS, WITH CITATIONS. Shipped UI under the Civ VI install's `steamassets/`
-- (this host: `~/.steam/debian-installation/steamapps/common/Sid Meier's Civilization VI/`):
--   * `base/assets/ui/endgame/endgamemenu.lua:1055-1069` -- `ShowEndGame`, the shipped path that
--     answers "what should this human be shown right now": `1057 if(player:IsAlive()) then`,
--     `1058 local victor, victoryType = Game.GetWinningTeam();`, `1059 if(victor ==
--     player:GetTeam()) then`, `1060 local victory = GameInfo.Victories[victoryType];`,
--     `1062 View(TeamVictoryData(victor, victory.VictoryType))`, and the fallthrough
--     `1069 View(PlayerDefeatedData(playerId, "DEFEAT_DEFAULT"))`. A dead `IsAlive()` is exactly
--     how the shipped UI decides to tell a human they lost.
--   * `endgamemenu.lua:868-875` -- the event-driven victory test, verbatim the same comparison:
--     `868 local localPlayerID = Game.GetLocalPlayer();` `870 local localPlayerTeamID =
--     pLocalPlayer:GetTeam();` `875 data.IsWinnerLocalPlayer = winningTeamID == localPlayerTeamID;`
--   * `endgamemenu.lua:925-926` -- the victory type's human name: `925 local victory =
--     GameInfo.Victories[victoryType];` `926 data.VictoryTypeHeader = victory.Name;`, rendered at
--     `636 Controls.VictoryTypeName:SetText(...)`. Rows: `base/assets/gameplay/data/victories.xml:33-39`
--     (plus `dlc/expansion2/data/expansion2_victories.xml:7` for `VICTORY_DIPLOMATIC`).
--   * `endgamemenu.lua:947-951` -- the winner's name: `948 local winnerPlayerID =
--     Teams[winningTeamID][1];` `951 data.WinnerName =
--     Locale.Lookup(pWinnerConfig:GetCivilizationDescription());`, rendered at
--     `637 Controls.VictoryPlayerName:SetText(...)`. The multiplayer-only " (PlayerName)" suffix
--     (`952-955`) is deliberately NOT reproduced: this harness plays single player, where the
--     screen does not show it.
--   * `endgamemenu.lua:832` / `969` -- when the local player was simply eliminated with nobody
--     having won, the screen's winner panel is empty (`data.WinnerName = ""`). That is why
--     `victory_type`/`winner` stay absent on an elimination defeat here.
--   * `base/assets/ui/worldinput.lua:3444-3445` -- the shipped "is the game over" probe,
--     `return Game.GetWinningTeam() ~= nil`, which is the authority for treating a nil return as
--     "nobody has won"; `base/assets/ui/menus/ingametopoptionsmenu.lua:357-359` tests it the same
--     way. `Game.GetWinningPlayer()` does NOT exist in the shipped UI (zero call sites anywhere
--     under `steamassets/`), so it is not used here.
--   * `base/assets/ui/menus/hotseatbackground.lua:12` --
--     `ContextPtr:LookUpControl("/InGame/EndGameMenu")` is how shipped code reaches the end-game
--     screen's own context; it is declared `base/assets/ui/ingame.xml:120` as
--     `<LuaContext ID="EndGameMenu" FileName="EndGameMenu" Hidden="1"/>` and queued at
--     `endgamemenu.lua:755 UIManager:QueuePopup(ContextPtr, PopupPriority.EndGameMenu)`.
--     Reported here as `end_game_screen_shown`, CORROBORATION ONLY -- never a gate. A defeat whose
--     popup failed to queue (the tutorial ruleset never subscribes at all:
--     `endgamemenu.lua:1327-1330`) is still a defeat, and gating on the screen would silently miss
--     it.
--
-- Expansion overrides change none of this: `dlc/expansion1/ui/replacements/
-- endgamemenu_expansion1.lua` and `dlc/expansion2/ui/replacements/endgamemenu_expansion2.lua`
-- only add the Historic Moments/Export buttons and a `Styles["VICTORY_DIPLOMATIC"]` entry.
--
-- SANDBOX CONSTRAINT (specs/002-civ-playing-harness/spikes/lua-api-verification-linux.md, P5):
-- neither tuner context exposes `require`, `io`, or `debug`, and no JSON library exists in either.
-- This file must stay entirely self-contained and carries its own JSON encoder.
--
-- UNVERIFIED LIVE: every accessor below is individually pcall'd and an unreadable value is simply
-- absent from the result -- never a fabricated outcome. `Players[id]:IsAlive()` was observed false
-- on the defeated player at 18:50 EDT; `Game.GetWinningTeam`, `GameInfo.Victories`, `Teams`,
-- `PlayerConfigurations:GetCivilizationDescription` and `Locale.Lookup` are the shipped screen's
-- own calls but have not been exercised from the InGame tuner state on a finished game. The Python
-- side treats any unreadable or unnameable result as "no game over" and plays the turn as usual.

local function CivSim_JsonEncode(value)
    local t = type(value)
    if value == nil then
        return "null"
    elseif t == "boolean" then
        return value and "true" or "false"
    elseif t == "number" then
        if value ~= value then return "null" end
        return tostring(value)
    elseif t == "string" then
        local escaped = value:gsub('[%c"\\]', function(c)
            if c == '"' then return '\\"'
            elseif c == '\\' then return '\\\\'
            elseif c == '\n' then return '\\n'
            elseif c == '\r' then return '\\r'
            elseif c == '\t' then return '\\t'
            else return string.format('\\u%04x', string.byte(c)) end
        end)
        return '"' .. escaped .. '"'
    elseif t == "table" then
        local n = 0
        for _ in pairs(value) do n = n + 1 end
        if n == 0 then return "[]" end
        local isArray = true
        for i = 1, n do if value[i] == nil then isArray = false break end end
        if isArray then
            local parts = {}
            for i = 1, n do parts[i] = CivSim_JsonEncode(value[i]) end
            return "[" .. table.concat(parts, ",") .. "]"
        else
            local parts = {}
            for k, v in pairs(value) do
                parts[#parts + 1] = CivSim_JsonEncode(tostring(k)) .. ":" .. CivSim_JsonEncode(v)
            end
            return "{" .. table.concat(parts, ",") .. "}"
        end
    else
        return "null"
    end
end

-- A displayable string, or nil. `Locale.Lookup` is applied exactly where the shipped screen
-- applies it (endgamemenu.lua:951 for the winner; the victory header is stored as its own LOC key
-- at :926 and resolved by the label, so it is resolved here too -- a raw `LOC_...` key is what a
-- reader of the record must never be handed as "what the human saw"). A lookup that fails falls
-- back to the key itself rather than to nothing: an untranslated name is still a name.
local function CivSim_GameOver_Localized(key)
    if type(key) ~= "string" or key == "" then return nil end
    local ok, text = pcall(function() return Locale.Lookup(key) end)
    if ok and type(text) == "string" and text ~= "" then return text end
    return key
end

-- endgamemenu.lua:925-926. `victoryType` is `Game.GetWinningTeam()`'s SECOND return value (:1058),
-- an index into `GameInfo.Victories`; the human-visible name is that row's `Name`.
local function CivSim_GameOver_VictoryName(victoryType)
    if victoryType == nil then return nil end
    local ok, row = pcall(function() return GameInfo.Victories[victoryType] end)
    if not ok or row == nil then return nil end
    local okName, name = pcall(function() return row.Name end)
    if not okName then return nil end
    return CivSim_GameOver_Localized(name)
end

-- endgamemenu.lua:948-951. The winning team's first member is the civilization the victory panel
-- names. Single player only, by construction: the " (PlayerName)" suffix at :952-955 is
-- multiplayer-only and is not reproduced.
local function CivSim_GameOver_WinnerName(winningTeam)
    local okTeam, members = pcall(function() return Teams[winningTeam] end)
    if not okTeam or members == nil then return nil end
    local okId, playerId = pcall(function() return members[1] end)
    if not okId or playerId == nil then return nil end
    local okCfg, config = pcall(function() return PlayerConfigurations[playerId] end)
    if not okCfg or config == nil then return nil end
    local okDesc, description = pcall(function() return config:GetCivilizationDescription() end)
    if not okDesc then return nil end
    return CivSim_GameOver_Localized(description)
end

-- hotseatbackground.lua:12-14 resolves the end-game screen's own context by name and asks the UI
-- manager about it; the harness's own established equivalent (lua/ingame/screens.lua) is
-- `ContextPtr:LookUpControl("/InGame/<State>")` plus `IsHidden()`, which is what is used here so
-- the two agree about what "shown" means. Reported, never decisive -- see the header.
local function CivSim_GameOver_EndScreenShown()
    local okCtx, context = pcall(function()
        return ContextPtr:LookUpControl("/InGame/EndGameMenu")
    end)
    if not okCtx or context == nil then return nil end
    local okHidden, hidden = pcall(function() return context:IsHidden() end)
    if not okHidden or type(hidden) ~= "boolean" then return nil end
    return hidden == false
end

local function CivSim_GameOver_Read()
    local result = { game_over = false }

    local localPlayer = nil
    local okLocal, lp = pcall(function() return Game.GetLocalPlayer() end)
    if okLocal then localPlayer = lp end
    if type(localPlayer) ~= "number" or localPlayer < 0 then
        -- No local player (an observer slot, or the game is not in a playable state). Reported
        -- honestly rather than guessed at; the Python side plays the turn as usual.
        result.reason = "local player is not resolvable from this state"
        return result
    end

    result.end_game_screen_shown = CivSim_GameOver_EndScreenShown()

    local alive = nil
    local okAlive, isAlive = pcall(function() return Players[localPlayer]:IsAlive() end)
    if okAlive and type(isAlive) == "boolean" then alive = isAlive end
    result.local_player_alive = alive

    local localTeam = nil
    local okTeam, team = pcall(function() return Players[localPlayer]:GetTeam() end)
    if okTeam then localTeam = team end

    -- Both return values (endgamemenu.lua:1058). A nil winning team is "nobody has won yet"
    -- (worldinput.lua:3444); the `>= 0` guard additionally covers a negative-id sentinel, so the
    -- only way this misreads is a NON-NEGATIVE sentinel for "no winner", which would misreport a
    -- live game as over. If a live client ever shows one, correct this guard.
    local winningTeam, victoryType = nil, nil
    local okWin, winner, winningVictory = pcall(function() return Game.GetWinningTeam() end)
    if okWin then
        winningTeam = winner
        victoryType = winningVictory
    end
    local haveWinner = (type(winningTeam) == "number" and winningTeam >= 0)

    if alive == false then
        -- endgamemenu.lua:1057/1069: not alive -> the human is shown the defeat screen, whatever
        -- else is true. This is the 2026-09-21 case exactly.
        result.game_over = true
        result.outcome = "defeat"
        result.basis = "local_player_not_alive"
    elseif haveWinner then
        result.game_over = true
        if localTeam ~= nil and winningTeam == localTeam then
            result.outcome = "victory"
            result.basis = "winning_team_is_local_team"
        else
            result.outcome = "defeat"
            result.basis = "winning_team_is_another_team"
        end
    else
        -- Alive, nobody has won: the game is still being played. Nothing below this line runs,
        -- which is the structural half of the parity gate in the header.
        return result
    end

    if haveWinner then
        result.victory_type = CivSim_GameOver_VictoryName(victoryType)
        result.winner = CivSim_GameOver_WinnerName(winningTeam)
    end
    return result
end

CivSim_GameOver = {
    read = CivSim_GameOver_Read,
}

-- Example dispatch (performed by run/game_over.py's LuaGameOverReader, not by this file):
--   print(CivSim_JsonEncode(CivSim_GameOver.read()))

File: src/civsim_harness/run/game_over.py
"""Game-over detection at the turn boundary (2026-09-21, gameplay).

MEASURED, 18:50 EDT 2026-09-21, ``run-…`` on the live Linux client: Persia was defeated at game
turn 59 (Georgia took the only city). The harness had no game-over detection at all, so the next
turn started exactly as any other: ``run/turn_cycle.py`` asked for a start-of-turn quicksave, the
game -- already sitting on ``EndGameMenu`` -- never wrote one, and the run paused with
``SaveVerificationError("expected .Civ6Save not found ... waited 10 s")``. A fresh ``InGame`` read
taken a minute later showed ``Players[0]:IsAlive() == false`` and ``EndGameMenu`` as the only shown
screen. The run's honest terminal state is a **defeat**, recorded as a stop resolution -- not a
save error dressed up as one, and not a ``paused`` run an operator has to go and interpret.

**Why this is a separate read rather than the existing ``game.outcome_state``.**
``catalogs/observations/game.yaml``'s ``game.outcome_state`` (T216, ``lua/gamecore/
game_outcome.lua``) already answers "has the local player won or lost", and ``run/composition.py``
feeds it to ``run/stop.py`` through ``evaluate_stop_facts``. But that value is read **inside** a
decision step's observation sweep and consumed **after** a turn finishes -- and the turn that
discovers the defeat is precisely the turn that cannot start, because its very first act
(FR-007's quicksave) is the thing the finished game refuses. The existing path is not wrong; it is
one turn too late to ever fire on this shape of ending. This module is the same question asked at
the one moment that answers it: before the quicksave, in the same pre-save position as the
blocking-prompt probe.

**Principle I -- what may be read, and when.** ``lua/gamecore/game_outcome.lua``'s parity note
forbids reporting *which* civilization won or *by which victory type* -- correctly, because while
the game is still being played that is opponent state a human cannot see. The moment the game is
over, the end-game screen states all of it to the human in plain text: the victory type's own
name, and the winning leader/civilization. So this module's Lua reads those two fields **only
once the game is already over**, from the same accessors ``endgamemenu.lua`` itself uses to build
that screen, and reports ``None`` for both at every other moment. That gate is in the Lua (see
``lua/ingame/game_over.lua``) and re-asserted here: :func:`interpret_game_over` drops
``victory_type``/``winner`` from any read that does not also say the game is over.

**A read that errors is not a game over.** Every failure mode -- a transport error, a malformed
body, a Lua accessor that does not exist on this build, an outcome string this module does not
recognise -- resolves to "no game over detected, proceed exactly as today". Recording a defeat
because a Lua read came back malformed would be strictly worse than the gap it closes; this is the
same rule ``run/composition.py``'s ``_interpret_game_outcome`` already states for its own path.
"""

from __future__ import annotations

from collections.abc import Awaitable, Callable, Mapping
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from civsim_harness.errors import HarnessError
from civsim_harness.models.run import StopResolution
from civsim_harness.run.stop import GameOutcome

#: Where ``lua/ingame/game_over.lua`` lives, relative to the repository root -- the same
#: ``lua_root``-relative convention every ``IntegrationCapability.implementation_ref`` uses
#: (``run/composition.py``'s ``DEFAULT_LUA_ROOT``), so one setting locates every Lua file the
#: harness ships whether or not a catalog declaration names it.
GAME_OVER_LUA_REF = "lua/ingame/game_over.lua"

#: The one dispatch-table entry ``lua/ingame/game_over.lua`` exposes.
_LUA_MODULE = "CivSim_GameOver"
_LUA_FUNCTION = "read"

GameOverReader = Callable[[], Awaitable[Any]]
"""The seam :class:`~civsim_harness.run.turn_cycle.TurnCycleDependencies` takes: a zero-argument
coroutine function returning whatever the game-over Lua printed (already JSON-decoded by
``NexusClient.execute_command``). Kept as a plain callable so ``run/turn_cycle.py`` has no
dependency on the transport, exactly as its ``read_observation_inputs``/``execute_action`` seams
do."""


@dataclass(frozen=True)
class GameOverRead:
    """One interpreted game-over read.

    ``game_over`` is the only field a caller should branch on: it is true **only** for a read that
    both says the game is over *and* names an outcome this module recognises. Everything else --
    a still-running game, an unreadable one, an outcome string from a future build -- lands with
    ``game_over=False``, which means "carry on with the turn", never "assume the worst".
    """

    game_over: bool
    local_player_alive: bool | None
    outcome: GameOutcome | None
    victory_type: str | None
    winner: str | None
    raw: Mapping[str, Any]

    @property
    def stop_resolution(self) -> StopResolution | None:
        """The :class:`~civsim_harness.models.run.StopResolution` this read terminates the run
        with, or ``None`` when it terminates nothing. Both values already exist in that enum
        (``victory``/``defeat``, data-model.md §4) -- nothing new is invented here."""
        if not self.game_over:
            return None
        if self.outcome is GameOutcome.VICTORY:
            return StopResolution.VICTORY
        if self.outcome is GameOutcome.DEFEAT:
            return StopResolution.DEFEAT
        return None

    def as_detail(self) -> dict[str, Any]:
        """The ``RunEvent.detail`` body for the ``game_over_detected`` event (Principle III).

        Carries the interpreted fields *and* the raw Lua result, so a later reader can tell a
        defeat the harness inferred from elimination apart from one it read off a rival's
        victory, without having to re-run anything.
        """
        return {
            "game_over": self.game_over,
            "local_player_alive": self.local_player_alive,
            "outcome": self.outcome.value if self.outcome is not None else None,
            "victory_type": self.victory_type,
            "winner": self.winner,
            "raw": dict(self.raw),
        }


#: The "nothing was detected" read, returned for every unreadable or unrecognised result.
NO_GAME_OVER = GameOverRead(
    game_over=False,
    local_player_alive=None,
    outcome=None,
    victory_type=None,
    winner=None,
    raw={},
)


def _optional_text(value: Any) -> str | None:
    """A displayable name, or ``None``. Empty and whitespace-only strings are ``None``: the
    end-game screen showing a blank line is not a name, and recording ``""`` as the winner would
    read as a fact rather than as the absence of one."""
    if not isinstance(value, str):
        return None
    stripped = value.strip()
    return stripped or None


def interpret_game_over(value: Any) -> GameOverRead:
    """Interpret one ``lua/ingame/game_over.lua`` result.

    Never raises: every shape that is not an explicit, recognised game over becomes
    :data:`NO_GAME_OVER`, which the turn cycle treats as "proceed exactly as today". In
    particular a ``game_over: true`` carrying an ``outcome`` this module does not recognise is
    **not** a game over here -- naming the ending is the whole point, and a run recorded as
    finished without one would violate ``Run``'s own "finished requires exactly one
    stop_resolution" invariant anyway.
    """
    if not isinstance(value, Mapping):
        return NO_GAME_OVER

    alive_raw = value.get("local_player_alive")
    local_player_alive = alive_raw if isinstance(alive_raw, bool) else None

    if value.get("game_over") is not True:
        return GameOverRead(
            game_over=False,
            local_player_alive=local_player_alive,
            outcome=None,
            victory_type=None,
            winner=None,
            raw=dict(value),
        )

    outcome_raw = value.get("outcome")
    if outcome_raw == GameOutcome.VICTORY.value:
        outcome = GameOutcome.VICTORY
    elif outcome_raw == GameOutcome.DEFEAT.value:
        outcome = GameOutcome.DEFEAT
    else:
        # "Over" with no outcome anyone here can name is not actionable: a run cannot be
        # `finished` without exactly one stop_resolution, so this is reported as no game over
        # rather than guessed at.
        return GameOverRead(
            game_over=False,
            local_player_alive=local_player_alive,
            outcome=None,
            victory_type=None,
            winner=None,
            raw=dict(value),
        )

    # Principle I: the victory type and the winner are exactly what the end-game screen states to
    # the human, and they are carried only on a read that says that screen's condition holds.
    return GameOverRead(
        game_over=True,
        local_player_alive=local_player_alive,
        outcome=outcome,
        victory_type=_optional_text(value.get("victory_type")),
        winner=_optional_text(value.get("winner")),
        raw=dict(value),
    )


class GameOverDetected(HarnessError):
    """The game was already over when this turn tried to begin (2026-09-21).

    Raised out of :func:`~civsim_harness.run.turn_cycle.run_turn_cycle` from the pre-save
    position, **after** the ``game_over_detected`` ``RunEvent`` is durably written and **before**
    the FR-007 quicksave is attempted -- because it cannot succeed: the client is showing
    ``EndGameMenu`` and will not write a save. Exactly like the headroom halt and the save failure
    beside it, "the turn never comes into existence as an attempt" (data-model.md §5's failure
    table), so no ``TurnCycle`` is written for it and the record carries no gap: a turn with no
    quicksave is not an attempted turn, so ``store/completeness.py`` never owes it a record.

    ``run/runner.py`` (T116) is what turns this into the run's terminal state: ``playing ->
    finished`` with :attr:`stop_resolution` -- the one recorded ``StopResolution`` (FR-005,
    invariant I10) -- rather than the ``paused`` every other unclassified mid-play ``HarnessError``
    lands in. A defeat is an ending, not a fault.
    """

    def __init__(
        self,
        message: str,
        *,
        stop_resolution: StopResolution,
        read: GameOverRead,
        detail: Mapping[str, Any] | None = None,
    ) -> None:
        super().__init__(message, detail=detail)
        self.stop_resolution = stop_resolution
        self.read = read


class LuaGameOverReader:
    """Dispatch ``lua/ingame/game_over.lua`` through an already-connected Nexus session.

    Built the same way ``saves/save_game.py``'s ``LuaSaveCapability`` is -- an
    ``execute(state_index, lua_body)`` callable plus a resolver for the current ``InGame`` state
    index, never a raw ``int``
    captured at construction (a Lua state index can differ across a phase transition, and a stale
    one silently runs the wrong Lua; see ``nexus/client.py``). The file's own source is sent
    verbatim with one appended ``print`` line, which is exactly what
    ``capability/executor.py`` does for every catalogued Lua file -- this reader exists only
    because this read deliberately carries **no** catalog declaration: it is harness bookkeeping at
    a turn boundary, never part of the agent's observation surface, and ``catalogs/observations/
    game.yaml`` is another lane's file today.
    """

    def __init__(
        self,
        execute: Callable[[int, str], Awaitable[Any]],
        *,
        in_game_state_index: Callable[[], int],
        lua_root: Path | str = Path("."),
    ) -> None:
        self._execute = execute
        self._in_game_state_index = in_game_state_index
        self._lua_root = Path(lua_root)
        self._source: str | None = None

    def _lua_body(self) -> str:
        if self._source is None:
            # Immutable within a catalog version, exactly as capability/executor.py assumes for
            # every other lua/**/*.lua file, so it is read once per reader.
            self._source = (self._lua_root / GAME_OVER_LUA_REF).read_text(encoding="utf-8")
        return (
            f"{self._source}\n\n"
            f"print(CivSim_JsonEncode({_LUA_MODULE}.{_LUA_FUNCTION}()))"
        )

    async def __call__(self) -> Any:
        return await self._execute(self._in_game_state_index(), self._lua_body())
